src/trace_segment.ml

```diff
@@ -0,0 +1,799 @@
+  if debug then print event time
+;;
+
+let add_event (t : t) (event : Event.Ok.Data.t) (time : Timestamp.t) =
+  print event time;
+  match event with
+  | Trace { trace_state_change = Some _; _ } -> ()
+  | Trace { kind = Some (Call | Syscall); src; dst; trace_state_change = _ } ->
+    handle_call t time ~src ~dst
+  | Trace { kind = Some Return; src; dst; trace_state_change = _ } ->
+    handle_return t time ~src ~dst
+  | Trace { kind = Some (Jump | Async); src; dst; trace_state_change = _ } ->
+    handle_jump t time ~src ~dst
+  | _ -> (* TODO *) ()
+;;
+
+let start_time t =
+  if Nonempty_vec.length t.callstacks = 1
+  then Null
+  else This (Nonempty_vec.get t.callstacks 1).#time
+;;
+
+let end_time t =
+  if Nonempty_vec.length t.callstacks = 1
+  then Null
+  else This (Nonempty_vec.last t.callstacks).#time
+;;
+
+let emit_frame_enter
+  (trace : Tracing.Trace.t)
+  thread
+  (time : Timestamp.t)
+  (location : Location.t)
+  =
+  if debug then Debug.eprintf "Enter %s\n" (Symbol.display_name location.symbol);
+  Tracing.Trace.write_duration_begin
+    trace (* TODO: populate arguments *)
+    ~args:[]
+    ~thread
+    ~name:(Symbol.display_name location.symbol)
+    ~time:(time :> Time_ns.Span.t)
+    ~category:""
+;;
+
+let emit_frame_exit
+  (trace : Tracing.Trace.t)
+  thread
+  (time : Timestamp.t)
+  (location : Location.t)
+  =
+  if debug then Debug.eprintf "Exit %s\n" (Symbol.display_name location.symbol);
+  Tracing.Trace.write_duration_end
+    trace
+    ~args:[]
+    ~thread
+    ~name:(Symbol.display_name location.symbol)
+    ~time:(time :> Time_ns.Span.t)
+    ~category:""
+;;
+
+let make_emit_trace_events trace thread = exclave_
+  Staged.stage (stack_ fun (#(prev, curr) : #(Callstack.t * Callstack.t)) ->
+    let[@inline always] emit_frame_enter time location =
+      emit_frame_enter trace thread time location
+    in
+    let[@inline always] emit_frame_exit time location =
+      emit_frame_exit trace thread time location
+    in
+    let time = curr.#time in
+    match curr.#control_flow with
+    | Jump ->
+      (* I'm not sure we even need to be recording a new callstack in the first place when
+         the symbol doesn't change, but I need to double-check. *)
+      if not (Symbol.equal prev.#leaf.location.symbol curr.#leaf.location.symbol)
+      then (
+        emit_frame_exit time prev.#leaf.location;
+        emit_frame_enter time curr.#leaf.location)
+    | Call ->
+      Frame.iter_until_rev curr.#leaf prev.#leaf.location.symbol ~f:(stack_ fun frame ->
+        emit_frame_enter time frame.location)
+      [@nontail]
+    | Return ->
+      Frame.iter_until prev.#leaf curr.#leaf.location.symbol ~f:(stack_ fun frame ->
+        emit_frame_exit time frame.location)
+      [@nontail])
+;;
+
+let write_trace
+  (t : t)
+  (trace : Tracing.Trace.t)
+  thread
+  ~enter_initial_callstack
+  ~exit_final_callstack
+  =
+  if Nonempty_vec.length t.callstacks > 1
```

Comment:
Why?

Reply:
We [start a new trace-segment each time we see encounter `trace_state_change = Some End`](https://github.com/Svetlitski/magic-trace/blob/9bbeaa5d48d5b33121b91f2abbf877d55014e2c2/src/trace_writer.ml#L1218-L1219), which has the consequence of making the very last trace-segment we ever create empty save for the sentinel callstack. It's also possible to end up here if we [encountered multiple decode errors in a row](https://github.com/Svetlitski/magic-trace/blob/9bbeaa5d48d5b33121b91f2abbf877d55014e2c2/src/trace_writer.ml#L1144).